src/sources/file/line_agg.rs

```diff
@@ -0,0 +1,519 @@
+where
+    T: Stream<Item = (Bytes, Filename), Error = ()>,
+{
+    /// `Bytes` - the message data; `Filename` - file name.
+    type Item = (Bytes, Filename);
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            // If we're in draining mode, short circut here.
+            if let Some(to_drain) = &mut self.draining {
+                if let Some((line, src)) = to_drain.pop() {
+                    return Ok(Async::Ready(Some((line, src))));
+                } else {
+                    return Ok(Async::Ready(None));
+                }
+            }
+
+            // Check for keys that have hit their timeout.
+            while let Ok(Async::Ready(Some(expired_key))) = self.timeouts.poll() {
+                self.expired.push_back(expired_key.into_inner());
+            }
+
+            match self.inner.poll() {
+                Ok(Async::Ready(Some((line, src)))) => {
+                    // Handle the incoming line we got from `inner`. If the
+                    // handler gave us something - return it, otherwise continue
+                    // with the flow.
+                    if let Some(val) = self.handle_line(line, src) {
+                        return Ok(Async::Ready(Some(val)));
+                    }
+                }
+                Ok(Async::Ready(None)) => {
+                    // We got `None`, this means the `inner` stream has ended.
+                    // Start flushing all existing data, stop polling `inner`.
+                    self.draining =
+                        Some(self.buffers.drain().map(|(k, v)| (v.into(), k)).collect());
+                }
+                Ok(Async::NotReady) => {
+                    // We didn't get any lines from `inner`, so we just give
+                    // a line from the expired lines queue.
+                    if let Some(key) = self.expired.pop_front() {
+                        if let Some(buffered) = self.buffers.remove(&key) {
+                            return Ok(Async::Ready(Some((buffered.freeze(), key))));
+                        }
+                    }
+
+                    return Ok(Async::NotReady);
+                }
+                Err(()) => return Err(()),
+            };
+        }
+    }
+}
+
+impl<T> LineAgg<T>
+where
+    T: Stream<Item = (Bytes, Filename), Error = ()>,
+{
+    /// Handle line, if we have something to output - return it.
+    fn handle_line(&mut self, line: Bytes, src: Filename) -> Option<(Bytes, Filename)> {
+        // Check if we already have the buffered data for the source.
+        if self.buffers.contains_key(&src) {
+            let condition_matched = self.config.condition_pattern.is_match(line.as_ref());
+            match self.config.mode {
+                // All consecutive lines matching this pattern are included in
+                // the group.
+                Mode::ContinueThrough => {
+                    if condition_matched {
```

Comment:
Did you consider doing this the other way around, with the `if condition_matched` outside a pair of `match self.config.mode`? This might make the double `get` usage I reference below easier to factor out.

Reply:
I didn’t. I can experiment with it, but overall we’re doing not just two, but four operations with HashMap - get/insert/remove. What could really help here is the `Entry` API, but one thing we beed from there, in particular, is not yet stable. I don’t know what our policy regarding unstable rust features is, so I just left it as is. I’d love to switch to `Entry` API though.

Reply:
I did some experiments, and I think the code is more readable if we group the logic for an operation mode together. It's not worth optimizing structure cause it'll hurt maintainability.

Reply:
Looks good, and it's always nice when something like this makes our code shorter to boot.

Reply:
Oops, commented on the wrong thread there. Agreed, readability is important. Branch prediction will likely deal with any potential performance consequences.

Reply:
Oh yes, and I'd expect compiler to rearrange code significantly anyway, so performance aspects are hardly something we should care about.